pkg/uefi/meregion.go

```diff
@@ -0,0 +1,228 @@
+// Copyright 2019 the LinuxBoot Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package uefi
+
+import (
+	"bytes"
+	"encoding/binary"
+	"encoding/hex"
+	"fmt"
+	"log"
+)
+
+// ME Partition parsing, the goal is to spot a padding in the ME Region
+// after the ME partitions so that this region can be shrunk in the IFD.
+//
+// ME partition informations from http://me.bios.io/ME_blob_format
+
+// MEFTPSignature is the sequence of bytes that an ME Flash Partition
+// Table is expected to start with.
+var (
+	MEFTPSignature = []byte{0x24, 0x46, 0x50, 0x54}
+)
+
+const (
+	// MEFTPSignatureLength represents the size of the ME FTP signature
+	MEFTPSignatureLength = 4
+	// MEPartitionTableEntryLength is the size of a partition table entry
+	MEPartitionTableEntryLength = 32
+)
+
+// MEFPT is the main structure that represents an ME Flash Partition Table.
+type MEFPT struct {
+	// Holds the raw buffer
+	buf []byte
+
+	PartitionCount    uint32
+	PartitionMapStart int
+	Entries           []MEPartitionEntry
+	// Metadata for extraction and recovery
+	ExtractPath string
+}
+
+// MEPartitionEntry is an entry in FTP
+type MEPartitionEntry struct {
+	Name     MEName
+	Owner    [4]byte
+	Offset   uint32
+	Length   uint32
+	Reserved [4]uint32
+}
+
+// MEName represent 4 bytes with JSON string support
+type MEName [4]byte
+
+// MarshalText converts MEName to a byte range (for JSON)
+func (n MEName) MarshalText() ([]byte, error) {
+	return bytes.TrimRight(n[:], "\x00"), nil
+}
+
+// UnmarshalText converts a byte range to MEName (for JSON)
+func (n *MEName) UnmarshalText(b []byte) error {
+	var m MEName
+	copy(m[:], b)
+	*n = m
+	if len(b) > len(m) {
+		return fmt.Errorf("Can’t unmarshal %q to MEName, %d > %d", b, len(b), len(m))
+	}
+	return nil
+}
+
+func (n MEName) String() string {
+	b, _ := n.MarshalText()
+	return string(b)
+}
+
+// FindMESignature searches for an Intel ME FTP signature
+func FindMESignature(buf []byte) (int, error) {
+	if bytes.Equal(buf[16:16+MEFTPSignatureLength], MEFTPSignature) {
+		// 16 + 4 since the descriptor starts after the signature
+		return 20, nil
```

Comment:
Return `16 + len(MEFTPSignature)` to be a bit more clear.

Reply:
Done